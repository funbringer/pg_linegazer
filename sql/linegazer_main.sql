\set VERBOSITY terse
SET search_path = 'public';

CREATE EXTENSION pg_linegazer;



/* Check htab drop */
select linegazer_clear();
select linegazer_clear();
select linegazer_clear();

/* Check wrong args */
select * from linegazer_simple_report(0);
select * from linegazer_simple_report(1);


CREATE OR REPLACE FUNCTION test_1()
RETURNS VOID AS
$$
declare a int4; begin a = 1; a = a + 2; a = a + 3; end;
$$ LANGUAGE plpgsql;


CREATE OR REPLACE FUNCTION test_2()
RETURNS VOID AS
$$
declare
	a int4;

begin
	a = 1;

	a = 2;

	a = 3;
end;
$$ LANGUAGE plpgsql;


CREATE OR REPLACE FUNCTION test_3()
RETURNS VOID AS
$$
declare
	a int4 := 0;
	i int4;

begin
	for i in 1..100 loop
		a = a + i;
	end loop;
end;
$$ LANGUAGE plpgsql;


CREATE OR REPLACE FUNCTION test_4()
RETURNS VOID AS
$$
declare
	a int4 := 0;
	i int4;

begin
	for i in 1..100 loop
		a = a + i;
	end loop;

	if i < 10 then
		raise exception 'wtf!';
	end if;
end;
$$ LANGUAGE plpgsql;


CREATE OR REPLACE FUNCTION test_5()
RETURNS VOID AS
$$
declare
	i int4;

begin
	for i in 1..4 loop
		perform test_2();

		if i % 2 = 0 then
			perform test_3();
		end if;
	end loop;
end;
$$ LANGUAGE plpgsql;


CREATE OR REPLACE FUNCTION test_6()
RETURNS VOID AS
$$
declare
	a int4;
	i int4;

begin
	for i in (select generate_series(1, 4)) loop
		a = a + i;

		if a > 100 then
			raise exception 'wtf!';
		end if;
	end loop;
end;
$$ LANGUAGE plpgsql;


CREATE OR REPLACE FUNCTION test_7()
RETURNS VOID AS
$$
declare
	a int4;
	i int4;

begin
	foreach i in array '{1,2,3,4}'::int4[] loop
		a = a + i;

		if a > 100 then
			raise exception 'wtf!';
		end if;
	end loop;
end;
$$ LANGUAGE plpgsql;


CREATE OR REPLACE FUNCTION test_8()
RETURNS VOID AS
$$
declare
	a int4;
	i int4;

begin
	for i in select * from generate_series(1, 2) loop
		a = a + i;

		if a > 100 then
			raise exception 'wtf!';
		end if;
	end loop;
end;
$$ LANGUAGE plpgsql;


CREATE OR REPLACE FUNCTION test_9()
RETURNS VOID AS
$$
declare
	a int4 := 0;

begin
	loop
		if a > 10 then
			exit;
		end if;

		if a > 100 then
			raise exception 'wtf!';
		end if;

		a = a + 1;
	end loop;
end;
$$ LANGUAGE plpgsql;


CREATE OR REPLACE FUNCTION test_10()
RETURNS VOID AS
$$
declare
	a int4 := 0;

begin
	while a < 10 loop

		if a > 100 then
			raise exception 'wtf!';
		elseif a > 1000 then
			raise exception 'wtf!';
		else
			a = a;
		end if;

		a = a + 1;
	end loop;
end;
$$ LANGUAGE plpgsql;


CREATE OR REPLACE FUNCTION test_11()
RETURNS VOID AS
$$
declare
	a int4 := 0;

begin
	a = 100;

	if a = 200 then
		raise exception 'wtf!';
	end if;

exception
	when division_by_zero then
		raise exception 'wtf!';
end;
$$ LANGUAGE plpgsql;


CREATE OR REPLACE FUNCTION test_12()
RETURNS VOID AS
$$
declare
	a int4;
	i int4;

begin
	for i in 1..10 loop
		case i
			when 1, 2, 3 then
				a = 1;
			when 4 then
				a = 2;
			when 100 then
				raise exception 'wtf!';
			else
				a = 3;
		end case;
	end loop;

end;
$$ LANGUAGE plpgsql;



/* Test function #1 */
select test_1();
select * from linegazer_simple_report('test_1'::regproc);
select test_1();
select * from linegazer_simple_report('test_1'::regproc);
select linegazer_clear();
select * from linegazer_simple_report('test_1'::regproc);


/* Test function #2 */
select test_2();
select * from linegazer_simple_report('test_2'::regproc);
select test_2();
select * from linegazer_simple_report('test_2'::regproc);
select linegazer_clear();
select * from linegazer_simple_report('test_2'::regproc);


/* Test function #3 */
select test_3();
select * from linegazer_simple_report('test_3'::regproc);
select test_3();
select * from linegazer_simple_report('test_3'::regproc);
select linegazer_clear();
select * from linegazer_simple_report('test_3'::regproc);


/* Test function #4 */
select test_4();
select * from linegazer_simple_report('test_4'::regproc);
select test_4();
select * from linegazer_simple_report('test_4'::regproc);
select linegazer_clear();
select * from linegazer_simple_report('test_4'::regproc);


/* Test function #5 */
select test_5();
select * from linegazer_simple_report('test_5'::regproc);
select * from linegazer_simple_report('test_2'::regproc);
select * from linegazer_simple_report('test_3'::regproc);
select test_5();
select * from linegazer_simple_report('test_5'::regproc);
select * from linegazer_simple_report('test_2'::regproc);
select * from linegazer_simple_report('test_3'::regproc);
select linegazer_clear();
select * from linegazer_simple_report('test_5'::regproc);
select * from linegazer_simple_report('test_2'::regproc);
select * from linegazer_simple_report('test_3'::regproc);


/* Test function #6 */
select test_6();
select * from linegazer_simple_report('test_6'::regproc);
select test_6();
select * from linegazer_simple_report('test_6'::regproc);
select linegazer_clear();
select * from linegazer_simple_report('test_6'::regproc);


/* Test function #7 */
select test_7();
select * from linegazer_simple_report('test_7'::regproc);
select test_7();
select * from linegazer_simple_report('test_7'::regproc);
select linegazer_clear();
select * from linegazer_simple_report('test_7'::regproc);


/* Test function #8 */
select test_8();
select * from linegazer_simple_report('test_8'::regproc);
select test_8();
select * from linegazer_simple_report('test_8'::regproc);
select linegazer_clear();
select * from linegazer_simple_report('test_8'::regproc);


/* Test function #9 */
select test_9();
select * from linegazer_simple_report('test_9'::regproc);
select test_9();
select * from linegazer_simple_report('test_9'::regproc);
select linegazer_clear();
select * from linegazer_simple_report('test_9'::regproc);


/* Test function #10 */
select test_10();
select * from linegazer_simple_report('test_10'::regproc);
select test_10();
select * from linegazer_simple_report('test_10'::regproc);
select linegazer_clear();
select * from linegazer_simple_report('test_10'::regproc);


/* Test function #11 */
select test_11();
select * from linegazer_simple_report('test_11'::regproc);
select test_11();
select * from linegazer_simple_report('test_11'::regproc);
select linegazer_clear();
select * from linegazer_simple_report('test_11'::regproc);


/* Test function #12 */
select test_12();
select * from linegazer_simple_report('test_12'::regproc);
select test_12();
select * from linegazer_simple_report('test_12'::regproc);
select linegazer_clear();
select * from linegazer_simple_report('test_12'::regproc);


/* Test native function */
select generate_series(1::int4, 1::int4);
select * from linegazer_simple_report('generate_series(int4,int4)'::regprocedure);
select linegazer_clear();


/* Test isolation */
select test_1();
select * from linegazer_simple_report('test_1'::regproc);
select test_2();
select * from linegazer_simple_report('test_1'::regproc);
select * from linegazer_simple_report('test_2'::regproc);
select test_3();
select * from linegazer_simple_report('test_1'::regproc);
select * from linegazer_simple_report('test_2'::regproc);
select * from linegazer_simple_report('test_3'::regproc);
select linegazer_clear();



DROP EXTENSION pg_linegazer;
