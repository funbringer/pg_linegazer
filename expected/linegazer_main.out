\set VERBOSITY terse
SET search_path = 'public';
CREATE EXTENSION pg_linegazer;
/* Check htab drop */
select linegazer_clear();
 linegazer_clear 
-----------------
 
(1 row)

select linegazer_clear();
 linegazer_clear 
-----------------
 
(1 row)

select linegazer_clear();
 linegazer_clear 
-----------------
 
(1 row)

/* Check wrong args */
select * from linegazer_simple_report(0);
ERROR:  wrong function 0
select * from linegazer_simple_report(1);
ERROR:  wrong function 1
CREATE OR REPLACE FUNCTION test_1()
RETURNS VOID AS
$$
declare a int4; begin a = 1; a = a + 2; a = a + 3; end;
$$ LANGUAGE plpgsql;
CREATE OR REPLACE FUNCTION test_2()
RETURNS VOID AS
$$
declare
	a int4;

begin
	a = 1;

	a = 2;

	a = 3;
end;
$$ LANGUAGE plpgsql;
CREATE OR REPLACE FUNCTION test_3()
RETURNS VOID AS
$$
declare
	a int4 := 0;
	i int4;

begin
	for i in 1..100 loop
		a = a + i;
	end loop;
end;
$$ LANGUAGE plpgsql;
CREATE OR REPLACE FUNCTION test_4()
RETURNS VOID AS
$$
declare
	a int4 := 0;
	i int4;

begin
	for i in 1..100 loop
		a = a + i;
	end loop;

	if i < 10 then
		raise exception 'wtf!';
	end if;
end;
$$ LANGUAGE plpgsql;
CREATE OR REPLACE FUNCTION test_5()
RETURNS VOID AS
$$
declare
	i int4;

begin
	for i in 1..4 loop
		perform test_2();

		if i % 2 = 0 then
			perform test_3();
		end if;
	end loop;
end;
$$ LANGUAGE plpgsql;
/* Test function #1 */
select test_1();
 test_1 
--------
 
(1 row)

select * from linegazer_simple_report('test_1'::regproc);
 line | hits |                          code                           
------+------+---------------------------------------------------------
    1 |      | 
    2 |    3 | declare a int4; begin a = 1; a = a + 2; a = a + 3; end;
(2 rows)

select test_1();
 test_1 
--------
 
(1 row)

select * from linegazer_simple_report('test_1'::regproc);
 line | hits |                          code                           
------+------+---------------------------------------------------------
    1 |      | 
    2 |    6 | declare a int4; begin a = 1; a = a + 2; a = a + 3; end;
(2 rows)

select linegazer_clear();
 linegazer_clear 
-----------------
 
(1 row)

select * from linegazer_simple_report('test_1'::regproc);
 line | hits |                          code                           
------+------+---------------------------------------------------------
    1 |    0 | 
    2 |    0 | declare a int4; begin a = 1; a = a + 2; a = a + 3; end;
(2 rows)

/* Test function #2 */
select test_2();
 test_2 
--------
 
(1 row)

select * from linegazer_simple_report('test_2'::regproc);
 line | hits |      code       
------+------+-----------------
    1 |      | 
    2 |      | declare
    3 |      |         a int4;
    4 |      | 
    5 |      | begin
    6 |    1 |         a = 1;
    7 |      | 
    8 |    1 |         a = 2;
    9 |      | 
   10 |    1 |         a = 3;
   11 |      | end;
(11 rows)

select test_2();
 test_2 
--------
 
(1 row)

select * from linegazer_simple_report('test_2'::regproc);
 line | hits |      code       
------+------+-----------------
    1 |      | 
    2 |      | declare
    3 |      |         a int4;
    4 |      | 
    5 |      | begin
    6 |    2 |         a = 1;
    7 |      | 
    8 |    2 |         a = 2;
    9 |      | 
   10 |    2 |         a = 3;
   11 |      | end;
(11 rows)

select linegazer_clear();
 linegazer_clear 
-----------------
 
(1 row)

select * from linegazer_simple_report('test_2'::regproc);
 line | hits |      code       
------+------+-----------------
    1 |    0 | 
    2 |    0 | declare
    3 |    0 |         a int4;
    4 |    0 | 
    5 |    0 | begin
    6 |    0 |         a = 1;
    7 |    0 | 
    8 |    0 |         a = 2;
    9 |    0 | 
   10 |    0 |         a = 3;
   11 |    0 | end;
(11 rows)

/* Test function #3 */
select test_3();
 test_3 
--------
 
(1 row)

select * from linegazer_simple_report('test_3'::regproc);
 line | hits |             code             
------+------+------------------------------
    1 |      | 
    2 |      | declare
    3 |      |         a int4 := 0;
    4 |      |         i int4;
    5 |      | 
    6 |      | begin
    7 |    1 |         for i in 1..100 loop
    8 |  100 |                 a = a + i;
    9 |      |         end loop;
   10 |      | end;
(10 rows)

select test_3();
 test_3 
--------
 
(1 row)

select * from linegazer_simple_report('test_3'::regproc);
 line | hits |             code             
------+------+------------------------------
    1 |      | 
    2 |      | declare
    3 |      |         a int4 := 0;
    4 |      |         i int4;
    5 |      | 
    6 |      | begin
    7 |    2 |         for i in 1..100 loop
    8 |  200 |                 a = a + i;
    9 |      |         end loop;
   10 |      | end;
(10 rows)

select linegazer_clear();
 linegazer_clear 
-----------------
 
(1 row)

select * from linegazer_simple_report('test_3'::regproc);
 line | hits |             code             
------+------+------------------------------
    1 |    0 | 
    2 |    0 | declare
    3 |    0 |         a int4 := 0;
    4 |    0 |         i int4;
    5 |    0 | 
    6 |    0 | begin
    7 |    0 |         for i in 1..100 loop
    8 |    0 |                 a = a + i;
    9 |    0 |         end loop;
   10 |    0 | end;
(10 rows)

/* Test function #4 */
select test_4();
 test_4 
--------
 
(1 row)

select * from linegazer_simple_report('test_4'::regproc);
 line | hits |                  code                   
------+------+-----------------------------------------
    1 |      | 
    2 |      | declare
    3 |      |         a int4 := 0;
    4 |      |         i int4;
    5 |      | 
    6 |      | begin
    7 |    1 |         for i in 1..100 loop
    8 |  100 |                 a = a + i;
    9 |      |         end loop;
   10 |      | 
   11 |    1 |         if i < 10 then
   12 |    0 |                 raise exception 'wtf!';
   13 |      |         end if;
   14 |      | end;
(14 rows)

select test_4();
 test_4 
--------
 
(1 row)

select * from linegazer_simple_report('test_4'::regproc);
 line | hits |                  code                   
------+------+-----------------------------------------
    1 |      | 
    2 |      | declare
    3 |      |         a int4 := 0;
    4 |      |         i int4;
    5 |      | 
    6 |      | begin
    7 |    2 |         for i in 1..100 loop
    8 |  200 |                 a = a + i;
    9 |      |         end loop;
   10 |      | 
   11 |    2 |         if i < 10 then
   12 |    0 |                 raise exception 'wtf!';
   13 |      |         end if;
   14 |      | end;
(14 rows)

select linegazer_clear();
 linegazer_clear 
-----------------
 
(1 row)

select * from linegazer_simple_report('test_4'::regproc);
 line | hits |                  code                   
------+------+-----------------------------------------
    1 |    0 | 
    2 |    0 | declare
    3 |    0 |         a int4 := 0;
    4 |    0 |         i int4;
    5 |    0 | 
    6 |    0 | begin
    7 |    0 |         for i in 1..100 loop
    8 |    0 |                 a = a + i;
    9 |    0 |         end loop;
   10 |    0 | 
   11 |    0 |         if i < 10 then
   12 |    0 |                 raise exception 'wtf!';
   13 |    0 |         end if;
   14 |    0 | end;
(14 rows)

/* Test function #5 */
select test_5();
 test_5 
--------
 
(1 row)

select * from linegazer_simple_report('test_5'::regproc);
 line | hits |                   code                    
------+------+-------------------------------------------
    1 |      | 
    2 |      | declare
    3 |      |         i int4;
    4 |      | 
    5 |      | begin
    6 |    1 |         for i in 1..4 loop
    7 |    4 |                 perform test_2();
    8 |      | 
    9 |    4 |                 if i % 2 = 0 then
   10 |    2 |                         perform test_3();
   11 |      |                 end if;
   12 |      |         end loop;
   13 |      | end;
(13 rows)

select * from linegazer_simple_report('test_2'::regproc);
 line | hits |      code       
------+------+-----------------
    1 |      | 
    2 |      | declare
    3 |      |         a int4;
    4 |      | 
    5 |      | begin
    6 |    4 |         a = 1;
    7 |      | 
    8 |    4 |         a = 2;
    9 |      | 
   10 |    4 |         a = 3;
   11 |      | end;
(11 rows)

select * from linegazer_simple_report('test_3'::regproc);
 line | hits |             code             
------+------+------------------------------
    1 |      | 
    2 |      | declare
    3 |      |         a int4 := 0;
    4 |      |         i int4;
    5 |      | 
    6 |      | begin
    7 |    2 |         for i in 1..100 loop
    8 |  200 |                 a = a + i;
    9 |      |         end loop;
   10 |      | end;
(10 rows)

select test_5();
 test_5 
--------
 
(1 row)

select * from linegazer_simple_report('test_5'::regproc);
 line | hits |                   code                    
------+------+-------------------------------------------
    1 |      | 
    2 |      | declare
    3 |      |         i int4;
    4 |      | 
    5 |      | begin
    6 |    2 |         for i in 1..4 loop
    7 |    8 |                 perform test_2();
    8 |      | 
    9 |    8 |                 if i % 2 = 0 then
   10 |    4 |                         perform test_3();
   11 |      |                 end if;
   12 |      |         end loop;
   13 |      | end;
(13 rows)

select * from linegazer_simple_report('test_2'::regproc);
 line | hits |      code       
------+------+-----------------
    1 |      | 
    2 |      | declare
    3 |      |         a int4;
    4 |      | 
    5 |      | begin
    6 |    8 |         a = 1;
    7 |      | 
    8 |    8 |         a = 2;
    9 |      | 
   10 |    8 |         a = 3;
   11 |      | end;
(11 rows)

select * from linegazer_simple_report('test_3'::regproc);
 line | hits |             code             
------+------+------------------------------
    1 |      | 
    2 |      | declare
    3 |      |         a int4 := 0;
    4 |      |         i int4;
    5 |      | 
    6 |      | begin
    7 |    4 |         for i in 1..100 loop
    8 |  400 |                 a = a + i;
    9 |      |         end loop;
   10 |      | end;
(10 rows)

select linegazer_clear();
 linegazer_clear 
-----------------
 
(1 row)

select * from linegazer_simple_report('test_5'::regproc);
 line | hits |                   code                    
------+------+-------------------------------------------
    1 |    0 | 
    2 |    0 | declare
    3 |    0 |         i int4;
    4 |    0 | 
    5 |    0 | begin
    6 |    0 |         for i in 1..4 loop
    7 |    0 |                 perform test_2();
    8 |    0 | 
    9 |    0 |                 if i % 2 = 0 then
   10 |    0 |                         perform test_3();
   11 |    0 |                 end if;
   12 |    0 |         end loop;
   13 |    0 | end;
(13 rows)

select * from linegazer_simple_report('test_2'::regproc);
 line | hits |      code       
------+------+-----------------
    1 |    0 | 
    2 |    0 | declare
    3 |    0 |         a int4;
    4 |    0 | 
    5 |    0 | begin
    6 |    0 |         a = 1;
    7 |    0 | 
    8 |    0 |         a = 2;
    9 |    0 | 
   10 |    0 |         a = 3;
   11 |    0 | end;
(11 rows)

select * from linegazer_simple_report('test_3'::regproc);
 line | hits |             code             
------+------+------------------------------
    1 |    0 | 
    2 |    0 | declare
    3 |    0 |         a int4 := 0;
    4 |    0 |         i int4;
    5 |    0 | 
    6 |    0 | begin
    7 |    0 |         for i in 1..100 loop
    8 |    0 |                 a = a + i;
    9 |    0 |         end loop;
   10 |    0 | end;
(10 rows)

/* Test native function */
select generate_series(1::int4, 1::int4);
 generate_series 
-----------------
               1
(1 row)

select * from linegazer_simple_report('generate_series(int4,int4)'::regprocedure);
 line | hits |         code         
------+------+----------------------
    1 |    0 | generate_series_int4
(1 row)

select linegazer_clear();
 linegazer_clear 
-----------------
 
(1 row)

/* Test isolation */
select test_1();
 test_1 
--------
 
(1 row)

select * from linegazer_simple_report('test_1'::regproc);
 line | hits |                          code                           
------+------+---------------------------------------------------------
    1 |      | 
    2 |    3 | declare a int4; begin a = 1; a = a + 2; a = a + 3; end;
(2 rows)

select test_2();
 test_2 
--------
 
(1 row)

select * from linegazer_simple_report('test_1'::regproc);
 line | hits |                          code                           
------+------+---------------------------------------------------------
    1 |      | 
    2 |    3 | declare a int4; begin a = 1; a = a + 2; a = a + 3; end;
(2 rows)

select * from linegazer_simple_report('test_2'::regproc);
 line | hits |      code       
------+------+-----------------
    1 |      | 
    2 |      | declare
    3 |      |         a int4;
    4 |      | 
    5 |      | begin
    6 |    1 |         a = 1;
    7 |      | 
    8 |    1 |         a = 2;
    9 |      | 
   10 |    1 |         a = 3;
   11 |      | end;
(11 rows)

select test_3();
 test_3 
--------
 
(1 row)

select * from linegazer_simple_report('test_1'::regproc);
 line | hits |                          code                           
------+------+---------------------------------------------------------
    1 |      | 
    2 |    3 | declare a int4; begin a = 1; a = a + 2; a = a + 3; end;
(2 rows)

select * from linegazer_simple_report('test_2'::regproc);
 line | hits |      code       
------+------+-----------------
    1 |      | 
    2 |      | declare
    3 |      |         a int4;
    4 |      | 
    5 |      | begin
    6 |    1 |         a = 1;
    7 |      | 
    8 |    1 |         a = 2;
    9 |      | 
   10 |    1 |         a = 3;
   11 |      | end;
(11 rows)

select * from linegazer_simple_report('test_3'::regproc);
 line | hits |             code             
------+------+------------------------------
    1 |      | 
    2 |      | declare
    3 |      |         a int4 := 0;
    4 |      |         i int4;
    5 |      | 
    6 |      | begin
    7 |    1 |         for i in 1..100 loop
    8 |  100 |                 a = a + i;
    9 |      |         end loop;
   10 |      | end;
(10 rows)

select linegazer_clear();
 linegazer_clear 
-----------------
 
(1 row)

DROP EXTENSION pg_linegazer;
